### PR TITLE
TDL-20675 Reduce API calls v1.5.2

### PR DESCRIPTION
**Pull request by prijendev**
# Description of change
- Added  support of config parameter `event_date_date_window`.
- Default `event_date_date_window` is 7 days and maximum possible value of it is 30 days.
- Make API call of `event_updates` for last 30 days only.
- If `start_date` or `last saved bookmark` value is before 30 days, then start the sync from the last 30 days only.
- Write maximum of replication key value or `sync_start_time - event_date_window` as bookmark for `event_updates`.
- Retry 429 error 7 times with exponential factor 2.

**Note**: We have removed assertion to compare fields values because of field discrepancy .
# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_stripe/__init__.py [hunk 1]
```diff
@@ -785,13 +804,19 @@ def sync_event_updates(stream_name):
...
         date_window_start = date_window_end
-        date_window_end = date_window_end + date_window_size
+        date_window_end = date_window_end + events_date_window_size
         singer.write_bookmark(Context.state,
                               stream_name + '_events',
                               'updates_created',
                               max_created)
         singer.write_state(Context.state)
 
+    max_created = max(max_created, sync_start_time - events_date_window_size)
```
Comment (savan-chovatiya) [inline tap_stripe/__init__.py:814]: This is correct logic though, can we keep it as start_time of the last date window as discussed?
```suggestion
    max_created = max(max_created, date_window_start )
```

File: tap_stripe/__init__.py [hunk 2]
```diff
@@ -97,6 +97,7 @@
 LOGGER = singer.get_logger()
 
 DEFAULT_DATE_WINDOW_SIZE = 30 #days
+DEFAULT_EVENT_DATE_WINDOW_SIZE = 7 #day
```
Comment (karanpanchal-crest) [inline tap_stripe/__init__.py:100]:
```suggestion
DEFAULT_EVENT_DATE_WINDOW_SIZE = 7 #days
```

File: tap_stripe/__init__.py [hunk 3]
```diff
@@ -106,6 +107,8 @@ class Context():
     stream_map = {}
     new_counts = {}
     updated_counts = {}
+    window_size = DEFAULT_DATE_WINDOW_SIZE
```
Comment (karanpanchal-crest) [inline tap_stripe/__init__.py:110]:
```suggestion
    default_start_date_window = DEFAULT_DATE_WINDOW_SIZE # By default start data collection from 30 days ago
```

File: tap_stripe/__init__.py [hunk 4]
```diff
@@ -106,6 +107,8 @@ class Context():
...
+    event_window_size = DEFAULT_EVENT_DATE_WINDOW_SIZE
```
Comment (karanpanchal-crest) [inline tap_stripe/__init__.py:111]:
```suggestion
    event_window_size = DEFAULT_EVENT_DATE_WINDOW_SIZE # By default collect data of 7 days in one API call
```

File: tap_stripe/__init__.py [hunk 5]
```diff
@@ -443,6 +446,11 @@ def sync_stream(stream_name):
         # when they are available via the API, so these streams will need
         # a short lookback window.
         if stream_name in IMMUTABLE_STREAMS:
+            if stream_name == "events":
+                # Start sync from 30 days before if bookmark/start_date is older than 30 days.
+                start_window = int(max(bookmark, (singer.utils.now() - timedelta(days=30)).timestamp()))
+                if start_window != bookmark:
+                    LOGGER.warning("Provided current bookmark/start_date is older than the last 30 days. So, starting sync for the last 30 days as Stripe Event API returns data for the last 30 days only.")
```
Comment (karanpanchal-crest) [inline tap_stripe/__init__.py:453]: @prijendev This line contains more than 120 characters. Did we make sure PEP8 standards are followed?
CC: @dbshah1212

Comment (dbshah1212) [reply tap_stripe/__init__.py:453]: No, As this isn't refactoring release. 
It's quick bug fix release. 
Tap-Stripe is a candidate of refactoring so we can plan this at the time of refactoring for sure.

File: tap_stripe/__init__.py (hunk 4 above)
Comment (prijendev) [reply tap_stripe/__init__.py]: Added comment as per suggestion.

File: tap_stripe/__init__.py [hunk 6]
```diff
@@ -106,6 +107,9 @@ class Context():
     stream_map = {}
     new_counts = {}
     updated_counts = {}
+    window_size = DEFAULT_DATE_WINDOW_SIZE # By default fetch data from last 30 days for newly created records.
+    event_update_window_size = DEFAULT_EVENT_UPDATE_DATE_WINDOW_SIZE # By default collect data of 7 days in one API
+                                                       # call for event_updates
```
Comment (dbshah1212) [inline tap_stripe/__init__.py:112]: This won't look good, I guess it will be better to take an exception or change the comment place.

**Review by dbshah1212 — APPROVED** [review]

**Review by savan-chovatiya — APPROVED** [review]

**Review by karanpanchal-crest — APPROVED** [review]

File: tap_stripe/__init__.py [hunk 7]
```diff
@@ -374,6 +380,23 @@ def reduce_foreign_keys(rec, stream_name):
                     rec['lines'][k] = [li.to_dict_recursive() for li in val]
     return rec
 
+# Retry 429 RateLimitError 7 times.
+@backoff.on_exception(backoff.expo,
+                        stripe.error.RateLimitError,
+                        max_tries=7,
+                        factor=2)
+def new_request(self, method, url, params=None, headers=None):
```
Comment (kethan1122) [inline tap_stripe/__init__.py:388]: can we name this function properly. **new_request** doesn't give any idea to reader or add comments

Comment (prijendev) [reply tap_stripe/__init__.py]: Renamed `new_rewuest` to `api_request` and added comments.

**Review by kethan1122 — APPROVED** [review]

File: tests/test_all_fields.py [hunk 8]
```diff
@@ -344,73 +387,4 @@ def all_fields_test(self, streams_to_test):
                 self.assertEqual(len(expected_pks_set), len(expected_pks))
 
                 # Verify by pks, that we replicated the expected records
-                self.assertTrue(actual_pks_set.issuperset(expected_pks_set))
```
Comment (prijendev) [inline tests/test_all_fields.py:347]: We have removed the assertion to compare field values because of field discrepancy due to the SDK  upgrade.

**Review by cosimon — COMMENTED** [review]
We have no tests that run the tap with the new config values. Please add them.

**Comment by prijendev** [comment]
> We have no tests that run the tap with the new config values. Please add them.

@cosimon Added  `event_date_window_size` in get_properties to run all tests with  newly added config param.

File: tests/test_event_updates.py [hunk 9]
```diff
@@ -1,16 +1,67 @@
...
+    """
+    @staticmethod
+    def name():
+        return "tt_stripe_event_sync_start"
+
+    def get_properties(self, *args):
+        """Configuration properties required for the tap."""
+
+        props = super().get_properties(*args)
+        props['event_date_window_size'] = 35
```
Comment (dbshah1212) [inline tests/test_event_updates.py:24]: Add comment

Comment (prijendev) [reply tests/test_event_updates.py]: Added comment for `event_date_window_size`.

**Comment by RushiT0122** [comment]
Tested following scenarios,

Unit tests -

```
- test_date_window_size
- test_logger_for_events
- test_rate_limit_error
- test_recursive_to_dict
- test_sync_event_updates
```

Integration Tests -

```
- test_discovery
- test_all_fields
- test_all_test_runs
- test_automatic_fields
- test_bookmarks
- test_create_object
- test_event_updates
- test_full_replication
- test_pagination
- test_start_date
```

**Pull request merged**